### PR TITLE
[Feature Restoration] repetition_penalty is essential for GLM-V models

### PR DESCRIPTION
**Pull request by zRzRzRzRzRzRzR**
PR #5703 attempted to add back this feature but was not merged. This PR fixes some inheritance class bugs that were incompatible with the current codebase based on that work, and adds MTP support for this feature.
For GLM-V benchmark testing and general usage, we recommend setting `repetition_penalty` to 1.1. Without this implementation, GLM-V responses tend to have higher repetition.

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_info.py [hunk 1]
```diff
@@ -294,6 +294,28 @@ def verify(
                 torch.repeat_interleave(linear_penalty, self.draft_token_num, dim=0)
             )
 
+            # Apply multiplicative penalties (repetition) directly on logits
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/eagle_info.py]: **Critical: Missing equivalent handling in `ngram_info.py`**

This explicit repetition penalty block is needed because the existing zero-tensor approach (`apply_logits_bias(zeros)`) makes multiplicative penalties a no-op (`0/penalty = 0`). The same zero-tensor pattern exists in `ngram_info.py:400-411`, but the PR does not add the equivalent explicit block there.

Result: `repetition_penalty` is **silently ignored** when using ngram speculative decoding — no error, no warning.

Please add the same explicit repetition penalty handling to `ngram_info.py`.

File: python/sglang/srt/speculative/eagle_info.py [hunk 2]
```diff
@@ -294,6 +294,28 @@ def verify(
...
+            import sglang.srt.sampling.penaltylib as penaltylib
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/eagle_info.py]: **Important: Move import to top of file**

This import is inside `verify()` which runs on every speculative verification step. There is no circular dependency — `penaltylib` does not import from `eagle_info`. All other imports in this file are top-level.

Move to the top-of-file imports instead.

File: python/sglang/srt/sampling/penaltylib/repetition_penalty.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        logits < 0,
+        logits * scaling_penalties,
+        logits / scaling_penalties,
+    )
+
+
+class BatchedRepetitionPenalizer(_BatchedPenalizer):
+    """
+    Repetition penalizer penalizes tokens based on their repetition in the input and output.
+    """
```
Comment (JustinTong0323) [inline python/sglang/srt/sampling/penaltylib/repetition_penalty.py:19]: **Important: Inaccurate docstring**

This says "in the input and output" but the implementation only tracks output tokens via `_cumulate_output_tokens`. Prompt (input) tokens are never marked in `cumulated_repetition_penalties`. This differs from the HuggingFace transformers convention where `repetition_penalty` applies to both prompt and generated tokens.

```suggestion
    Repetition penalizer penalizes tokens based on their presence in the generated output.
```

File: python/sglang/srt/speculative/eagle_info.py [hunk 4]
```diff
@@ -294,6 +294,28 @@ def verify(
...
+                        rep_pen.cumulated_repetition_penalties,
+                        self.draft_token_num,
+                        dim=0,
+                    )
+                    mask = expanded_penalties != 1.0
+                    if mask.any():
+                        logits = logits_output.next_token_logits
+                        pos = (logits > 0) & mask
+                        neg = (logits <= 0) & mask
+                        logits[pos] = logits[pos] / expanded_penalties[pos]
```
Comment (JustinTong0323) [inline python/sglang/srt/speculative/eagle_info.py]: **Important: Duplicated logic with subtle inconsistency**

This re-implements the scaling formula from `apply_scaling_penalties` instead of reusing it. The boundary condition differs:
- Here: `logits > 0` (divide) and `logits <= 0` (multiply) — zero goes to multiply branch
- `apply_scaling_penalties`: `logits < 0` (multiply), else divide — zero goes to divide branch

Both produce `0` for zero-valued logits, so it is functionally equivalent, but the inconsistency is a maintenance hazard. Consider refactoring to reuse `apply_scaling_penalties`, or at minimum align the boundary.

File: python/sglang/srt/sampling/sampling_batch_info.py [hunk 5]
```diff
@@ -228,15 +233,31 @@ def update_penalties(self):
                 dtype=torch.float32,
                 device=self.temperatures.device,
             )
-            self.penalizer_orchestrator.apply(self.acc_linear_penalties)
+            self.acc_scaling_penalties = None
+
+            for pen in self.penalizer_orchestrator.penalizers.values():
+                if not pen._is_prepared:
+                    continue
+                if isinstance(pen, penaltylib.BatchedRepetitionPenalizer):
```
Comment (JustinTong0323) [inline python/sglang/srt/sampling/sampling_batch_info.py]: **Suggestion: `isinstance` check is fragile**

This special-cases `BatchedRepetitionPenalizer` by type. If a future multiplicative penalizer is added, it would silently be treated as additive (`pen.apply(acc_linear_penalties)` on a zeros tensor → no effect).

Consider adding an `is_multiplicative` property to the `_BatchedPenalizer` base class, or at minimum add a comment here warning that any new multiplicative penalizer must be handled in this block.

**Review by JustinTong0323 — COMMENTED** [review]
## Review Summary

**+124/-3** across 5 files. Restores `repetition_penalty` support with a new `BatchedRepetitionPenalizer`, overlap-mode integration, and EAGLE speculative decoding support.

### Critical
1. **Repetition penalty silently ignored in ngram speculative decoding** — `ngram_info.py` uses the same zero-tensor pattern as eagle but has no explicit repetition penalty block. Multiplicative penalty on zeros is a no-op → penalty completely lost. (See inline comment on `eagle_info.py`)
2. **`repetition_penalty=0.0` causes division by zero** — `sampling_params.py:127` allows `0.0 <= repetition_penalty`, but `apply_scaling_penalties` divides positive logits by the penalty → produces `inf`/`nan`. Fix: change to `0.0 < self.repetition_penalty <= 2.0`.

### Important
3. Inline import in hot path (`eagle_info.py`) — see inline
4. Misleading docstring claims input+output penalty (`repetition_penalty.py`) — see inline
5. Duplicated penalty logic with boundary inconsistency (`eagle_info.py`) — see inline

### Suggestions
6. **No tests included.** Existing `test_penaltylib.py` has a clear template for all other penalizers. `TestUpdatePenalties` mock test in `test_sampling_batch_info.py` will likely break since the overlap-mode code no longer calls `orchestrator.apply()` directly.
7. `isinstance` special-casing in `update_penalties` is fragile for future multiplicative penalizers — see inline

**Review by hnyls2002 — CHANGES_REQUESTED** [review]
This PR bundles three independent changes into one:

1. **`repetition_penalty` restoration** — new `BatchedRepetitionPenalizer`, overlap mode and speculative decoding integration. This is a **general-purpose feature**, not GLM-V specific.
2. **MRoPE interleaved axis mapping fix** — fixes axis assignment for uneven `mrope_section` like `[8,12,12]`. GLM-V specific.
3. **`glm4_moe.py` FP32 gate projection cast** — numerical stability fix for MoE routing. GLM-V specific.

These touch completely separate subsystems (sampling, rotary embedding, MoE routing) with no code dependency between them. Please split into 2-3 PRs:

- One for `repetition_penalty` (sampling + speculative decoding changes)
- One for MRoPE axis mapping fix
- FP32 cast can go with either GLM-V PR or standalone

This makes review, bisect, and potential revert much easier.

---

Additionally, there is a correctness concern in the speculative decoding path (both `eagle_info.py` and `ngram_info.py`):

`sampling_info.apply_logits_bias(linear_penalty)` calls `orchestrator.apply()` which iterates **all** penalizers including the new `BatchedRepetitionPenalizer`. The multiplicative penalty (`where(logit < 0, logit * p, logit / p)`) should not be applied to the capture tensor — it can corrupt additive penalty values already accumulated by frequency/presence penalizers depending on dict iteration order. The manual repetition penalty application on actual logits (lines 301-315) is correct, but the implicit one via `orchestrator.apply()` on the capture tensor is not.

Suggested fix: in the speculative decoding verify paths, replace `sampling_info.apply_logits_bias(linear_penalty)` with a loop that only applies non-multiplicative penalizers + logit_bias to the capture tensor.

**Review by hnyls2002 — APPROVED** [review]
